### PR TITLE
Evaluation metrics of a top-k recommender 

### LINKED ISSUE #175: Evaluation - Item Coverage metric
Implement the `ItemCoverage` metric, that would receive in the constructor the `num_unique_items` and a cut-off `k`.
It should keep a fixed sized int tensor with `shape = (num_unique_items)` holding the count of items frequencies over batches, but only for the top-k predicted scores. The counts are not really needed (only for debugging maybe), it could also be just a bool tensor.
The metric result would be the something like `tf.cast(item_frequencies > 0, tf.float32) / tf.shape(item_frequencies)[0]`, i.e., the percentage of the item catalog that was recommended at least once among top-k for any example.

### PR DESCRIPTION
**Pull request by sararb**
Fixes #175 #167 

### Goals :soccer:
- Implement evaluation metrics for top-k recommender models: PopularityBias, Novelty, and Item coverage. 

### Implementation Details :construction:
- Add `get_candidate_probs` as tf util function to transform item frequencies into a probability distribution 
- Add custom Keras metrics to compute: popularity bias, novelty, and coverage from top-k predictions returned by a recommendation model. 

### Testing Details :mag:
- Add unit tests that check the scores returned by the implemented metrics.
- Update `test_topk_index` with evaluation metrics calculation.

### CODE REVIEW

File: merlin/models/tf/metrics/evaluation.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+def novelty_at(
+    predicted_candidates_probs: tf.Tensor,
+    k: int = 5,
+) -> tf.Tensor:
+    """
+    Computes novely@K metric
+    ----------
+    {METRIC_PARAMETERS_DOCSTRING}
+    """
+    return -tf.math.log(predicted_candidates_probs[:, :k] + MIN_FLOAT)
```
Comment (gabrielspmoreira) [inline merlin/models/tf/metrics/evaluation.py:50]: I think we should rename `MIN_FLOAT` to `EPSILON`, as there is another `MIN_FLOAT` constant in `merlin.models.utils.constants` used to downscore the false negatives

File: merlin/models/tf/metrics/evaluation.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+        candidate_probs = get_candidate_probs(item_freq_probs, self.is_prob_distribution)
+        self.candidate_probs = tf.Variable(
+            candidate_probs,
+            name="candidate_probs",
+            trainable=False,
+            dtype=tf.float32,
+            validate_shape=False,
+            shape=tf.shape(candidate_probs),
+        )
```
Comment (gabrielspmoreira) [inline merlin/models/tf/metrics/evaluation.py:110]: I think we could have a method `PopularityMetric.update_candidate_probs()` like [here](https://github.com/NVIDIA-Merlin/models/blob/main/merlin/models/tf/blocks/core/transformations.py#L673), to allow for users being able to update the probs.
That could be useful for example if training on streaming data, where the items popularity could be adjusted over time

File: merlin/models/tf/metrics/evaluation.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+def popularity_bias_at(
+    predicted_candidates_probs: tf.Tensor,
+    k: int = 5,
+) -> tf.Tensor:
+    """
+    Computes popularity bias@K metric
+    ----------
+    {METRIC_PARAMETERS_DOCSTRING}
+    """
+    return tf.reduce_sum(predicted_candidates_probs[:, :k], -1) / k
```
Comment (gabrielspmoreira) [inline merlin/models/tf/metrics/evaluation.py:62]: I think `Mean` base class expects metric results to be 1D (batch size) or 2D (batch size x (top-k) values per example).
You can see that in [recall_at()](https://github.com/NVIDIA-Merlin/models/blob/main/merlin/models/tf/metrics/ranking.py#L60-L66) we return a 1D tensor. That, way, the base class can multiply the metrics per example by `sample_weight` (from `update_state()`), before computing the (weighted) average.
So I think here we should just return either `tf.reduce_mean(predicted_candidates_probs[:, :k], -1)` or `predicted_candidates_probs[:, :k]`

File: merlin/models/tf/metrics/evaluation.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        The prediction coverage of the recommendations at the k-th position.
+    """
+
+    def __init__(self, num_unique_items: int, k=5, name: str = None, dtype=None, **kwargs):
+        if name is not None:
+            name = f"{name}_{k}"
+        super().__init__(name=name, dtype=dtype)
+
+        self.k = k
+        self.num_unique_items = num_unique_items
```
Comment (gabrielspmoreira) [inline merlin/models/tf/metrics/evaluation.py:220]: Maybe we could make `self.num_unique_items` a scalar non-trainable variable, and have an `update()` method to allow for users to update `num_unique_items` (with variable assign), as that might be useful when training models with streaming data, as the (number of )available items at a given point of time might change depending on the domain (e.g. on news recommendation)

File: tests/tf/metrics/test_metrics_popularity.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    predictions = tf.convert_to_tensor(
+        [[1, 9, 8, 7, 6], [1, 4, 3, 2, 5], [5, 9, 8, 7, 6]], tf.int64
+    )
+    item_freq_probs = [0.1, 0.05, 0.2, 0.05, 0.1, 0.1, 0.15, 0.05, 0.1, 0.1]
+    return labels, predictions, item_freq_probs
+
+
+def test_novelty_at_k(popularity_metrics_test_data):
+    labels, predictions, item_freq_probs = popularity_metrics_test_data
+    metric1 = NoveltyAt(item_freq_probs=item_freq_probs, is_prob_distribution=True, k=3)
```
Comment (gabrielspmoreira) [inline tests/tf/metrics/test_metrics_popularity.py:35]: Maybe we should have `is_prob_distribution` as a test argument and test both scenarios (True and False)

**Review by gabrielspmoreira — APPROVED** [review]
The metrics cannot currently be included in the `PredictionTask` because they require to receive in `update_state()` the predicted item ids instead of the prediction scores (y_pred) and labels (y_true), as usual.
I have tested the metrics locally following the example from the unit test and they do work properly when been called after `model.fit()` and after exporting to topk recommender. So I am merging this PR to be able to use them for the experiments.
I have created an issue #372 to have those metrics compatible with the other metrics, which are available for training and evaluation.
